osu.Game/Beatmaps/FramedBeatmapClock.cs

```diff
@@ -0,0 +1,214 @@
+            }
+            else
+            {
+                finalClockSource = decoupledClock;
+            }
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            if (applyOffsets)
+            {
+                Debug.Assert(userBeatmapOffsetClock != null);
+                Debug.Assert(userGlobalOffsetClock != null);
+
+                userAudioOffset = config.GetBindable<double>(OsuSetting.AudioOffset);
+                userAudioOffset.BindValueChanged(offset => userGlobalOffsetClock.Offset = offset.NewValue, true);
+
+                beatmapOffsetSubscription = realm.SubscribeToPropertyChanged(
+                    r => r.Find<BeatmapInfo>(beatmap.Value.BeatmapInfo.ID)?.UserSettings,
+                    settings => settings.Offset,
+                    val =>
+                    {
+                        userBeatmapOffsetClock.Offset = val;
+                    });
+            }
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+            finalClockSource.ProcessFrame();
+        }
+
+        private double totalAppliedOffset
+        {
+            get
+            {
+                if (!applyOffsets)
+                    return 0;
+
+                Debug.Assert(userGlobalOffsetClock != null);
+                Debug.Assert(userBeatmapOffsetClock != null);
+                Debug.Assert(platformOffsetClock != null);
+
+                return userGlobalOffsetClock.RateAdjustedOffset + userBeatmapOffsetClock.RateAdjustedOffset + platformOffsetClock.RateAdjustedOffset;
+            }
+        }
+
+        #region Delegation of IAdjustableClock / ISourceChangeableClock to decoupled clock.
+
+        public void ChangeSource(IClock? source)
+        {
+            Track = source as Track;
+            decoupledClock.ChangeSource(source);
+        }
+
+        public IClock? Source => decoupledClock.Source;
+
+        public void Reset()
+        {
+            decoupledClock.Reset();
+            finalClockSource.ProcessFrame();
+        }
+
+        public void Start()
+        {
+            decoupledClock.Start();
+            finalClockSource.ProcessFrame();
+        }
+
+        public void Stop()
+        {
+            decoupledClock.Stop();
+            finalClockSource.ProcessFrame();
+        }
+
+        public bool Seek(double position)
+        {
+            bool success = decoupledClock.Seek(position - totalAppliedOffset);
+            finalClockSource.ProcessFrame();
+
+            return success;
+        }
+
+        public void ResetSpeedAdjustments() => decoupledClock.ResetSpeedAdjustments();
+
+        public double Rate
+        {
+            get => decoupledClock.Rate;
+            set => decoupledClock.Rate = value;
+        }
+
+        #endregion
+
+        #region Delegation of IFrameBasedClock to clock with all offsets applied
+
+        public double CurrentTime => finalClockSource.CurrentTime;
```

Comment:
I have a feeling that the fact that `Seek()` operates on pre-offset time while `CurrentTime` returns post-offset time is going to be a cause of confusion for someone at some point. Willing to go along with it for now, but maybe with more xmldoc (the class's xmldoc mentions this somewhat, but only the seek half).

Reply:
Yeah, was tempted to rename `Seek`. I guess the important thing to note is that as a gameplay clock we are generally seeking to beatmap-provided times, but the offset being applied is the user expectation of where the game/beatmap playback should end up.

Reply:
As far as I understand, the idea is that the gameplay clock will always reflect whatever time inputted in `Seek`, but internally the track is shifted away from that time based on the offset applied by the user. So I'm slightly failing to see how it's going to be a cause of confusion, and the concept of "pre-offset time" and "post-offset time" feels off to grasp (if `CurrentTime` returns post-offset time, then `Seek` operates on a post-offset time as well?).

Reply:
That's the precise issue here: `CurrentTime` is post and `Seek` is pre-offset.

If `Seek` was switched to post-offset, every call to seek would have to account for this locally.